src/ValidatorChain.php

```diff
@@ -69,9 +69,8 @@ private function attachValidator($event, $callback, $priority)
             array_unshift($callback, $test);
         }
         if ($context instanceof ValidatorInterface) {
-            $data = $context->getData();
             $name = $context->getName();
-            $this->getStorage()->setMetadata('_VALID', [$name => $data]);
+            $this->getStorage()->setMetadata('_VALID', [$name => '']);
```

Comment:
The initial validator state has to be persisted somewhere. It looks like in the current implementation, each validator is serialised as `['ValidatorName' => 'Whatever $data is']`

My suggestion would be to re-architect this so that an `Environment` value object is added to the session metadata when it is first initialised. Then, prior to validation, the necessary validators are constructed with
```php
new WhateverValidator(
    Environment::fromSessionMetaData(), // Or whatever naming makes sense
    Environment::fromRequestOrGlobals(),
);
```